Add Redis helper library

pandah3 opened the pull request:
Fixes #110 

Adding a helper library that Node apps can use to instantiate a Redis client compatible with VIP Go.
_Goal:_ Make it easier for clients to use Redis with VIP Go configurations.

### TESTING
1. Run `redis-server`
2. Run `REDIS_MASTER=[HOST]:[PORT] node ./src/redis/index.js`
3. Should log message `Connected to Redis client`

mjangda commented:
@pandah3 would be good to think about and share how this can be tested locally.

pandah3 commented:
Yup, working on that now!

elazzabi commented:
pushed a small fix for `eslint`, still not sure why it was failing 🤷‍♂ 

something I saw is there is no way for clients to use their own logger in this. Is there a way to pass a logger when initiating the helper?

elazzabi commented on src/redis/index.js:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Internal dependencies
+ */
+let { logger } = require( '../logger/index' );
```
Given the user will pass a logger in line #L80 (or we'll default to `console`) we probably can remove this altogether and rely on the object passed. right?

pandah3 replied:
When I removed that line, I got `logger is undefined` errors because I wouldn't be defining the logger until the client is exported.

Would it be alright if I just:

`let logger = console;`
instead?

elazzabi replied:
I guess given you pass the logger in the constructor, it should be present in `this.logger` ?

Not sure if you need to explicitly set it in `this.logger` or it will be done automatically, but usually this is how it should work in a class.

elazzabi commented on src/redis/index.js:
```diff
@@ -0,0 +1,80 @@
+// Create a class in case we need to scale the connections
+class Redis {
+	constructor() {
+		const [ host, port ] = ( process.env.REDIS_MASTER || '' ).split( ':' );
```
I think this needs to move inside the `else`, i.e. after we check if `process.env.REDIS_MASTER` is already there.

pandah3 replied:
Ah right, moved 👍

pandah3 commented:
Fixed the changed suggestions 🙏 

Also added a `README` file.

Trying to see if I can stop the never-ending reconnection errors when Redis can't connect to a server, but it seems like it just stops and stalls..

elazzabi commented on src/redis/README.md:
````diff
@@ -0,0 +1,27 @@
+const { redis } = require( '@automattic/vip-go' );
+const redis = new Redis();
+```
+By default, this helper library logs to the console. Pass in your own logger like so:
````
```suggestion
By default, this helper library logs to the `console`. If you prefer, you can set your own logger:
```

pandah3 replied:
Changed 👍

elazzabi commented on src/redis/README.md:
````diff
@@ -0,0 +1,27 @@
+By default, this helper library logs to the console. Pass in your own logger like so:
+
+```js
+redis( { logger: yourLogger } );
````
```suggestion
const redis = new Redis( { logger: yourLogger } );
```

pandah3 replied:
🤦‍♀duh. Thank you!

elazzabi commented on src/redis/README.md:
```diff
@@ -0,0 +1,27 @@
+## Reconnecting to Redis
+If the connection to the Redis server is lost, offline queuing for commands will be disabled.
+
+After x amount of reconnection attempts, all pending commands will be flushed so that there are no stale commands.
```
```suggestion
After 3 reconnection attempts, all pending commands will be flushed so that there are no stale commands.
```

pandah3 replied:
Updated 👍

elazzabi commented on src/redis/README.md:
```diff
@@ -0,0 +1,27 @@
+If the connection to the Redis server is lost, offline queuing for commands will be disabled.
+
+After x amount of reconnection attempts, all pending commands will be flushed so that there are no stale commands.
+You may set the number of attempts with the `QUEUED_CONNECTION_ATTEMPTS` environment variable, or it will default to three reconnection attempts.
```
```suggestion
You can change this number by using the `QUEUED_CONNECTION_ATTEMPTS` environment variable.
```

pandah3 replied:
Changed 👍

elazzabi commented:
Yay thank you @pandah3 for working on this and adding the README 🙌 

I left some notes and some suggestions on the README

elazzabi commented on src/redis/README.md:
````diff
@@ -0,0 +1,27 @@
+Instantiate a new Redis client:
+
+``` js
+const { redis } = require( '@automattic/vip-go' );
````
I think we still need to expose `redis` in https://github.com/Automattic/vip-go-node/blob/master/src/index.js in order to be able to import it like this.

In addition, I see we are exporting a `new Redis()` here https://github.com/Automattic/vip-go-node/pull/125/files#diff-66b52dd10fa44ca40f44c06793a2aec1R77 but I guess in order to give the client the possibility to pass their own logger, we should just export the `Redis` class, and they need to do the `new Redis ( { logger: myLogger } );` on their own.

Or if we can avoid exporting classes, that would be good as all our other helpers (logger, newrelic, server) do not use a class approach when used by clients. It would be good to keep a consistency of:

```
const { logger, server, newrelic, redis } = require( '@automattic/vip-go-node' );

logger();
server();
newrelic();
redis();
```

We can use the class approach in the helper itself, but just exposing a function instead to keep the consistency. Hope I've been clear, if it's not the case, please let me know

elazzabi commented:
Whoa tried to rebase this on master, but I guess I somehow 💥 everything 👀

elazzabi commented:
This was continued in https://github.com/Automattic/vip-go-node/pull/190.

Will close this PR
